src/cc_catalog_airflow/dags/util/loader/sql.py

```diff
@@ -390,3 +391,59 @@ def update_flickr_sub_providers(
     Drop the temporary table
     """
     postgres.run(f'DROP TABLE public.{temp_table};')
+
+
+def update_europeana_sub_providers(
+  postgres_conn_id,
+  image_table=IMAGE_TABLE_NAME,
+  default_provider=prov.EUROPEANA_DEFAULT_PROVIDER,
+  sub_providers=prov.EUROPEANA_SUB_PROVIDERS
+):
+    postgres = PostgresHook(postgres_conn_id=postgres_conn_id)
+
+    select_query = dedent(
+        f'''
+        SELECT
+        {col.FOREIGN_ID} AS foreign_id,
+        {col.META_DATA} ->> 'dataProvider' AS data_providers
+        FROM {image_table}
+        WHERE {col.PROVIDER} = '{default_provider}';
+        '''
+    )
+
+    selected_records = postgres.get_records(select_query)
+
+    """
+    Filter the records to retain the ones with desired sub providers.
+    Each record in the new list would contain the sub provider name at the end.
+    """
+    filtered_records = []
+    for row in selected_records:
+        data_providers = ast.literal_eval(row[1])
+
+        eligible_sub_providers = {s for s in sub_providers if sub_providers[s]
+                                  in data_providers}
+        if len(eligible_sub_providers) > 1:
+            raise Exception(f"More than one sub-provider identified for the "
+                            f"image with foreign ID {row[0]}")
+        source = eligible_sub_providers.pop() if \
+            len(eligible_sub_providers) == 1 else None
+
+        if source is not None:
+            filtered_records.append(list(row) + [source])
+
+    for row in filtered_records:
+        foreign_id = row[0]
+        sub_provider = row[2]
+        postgres.run(
+            dedent(
+                f'''
+                UPDATE {image_table}
+                SET {col.SOURCE} = '{sub_provider}'
```

Comment:
I can probably do the update in the first for loop itself rather than filtering into a new `filtered_records` list.

Reply:
That would be preferable to save memory consumption.